// @tanstack/angular-query-experimental/build/esm2022/providers.mjs
import { DestroyRef, ENVIRONMENT_INITIALIZER, Injector, PLATFORM_ID, computed, effect, inject, makeEnvironmentProviders, runInInjectionContext, } from '@angular/core';
import { onlineManager } from '@tanstack/query-core';
import { DOCUMENT, isPlatformBrowser } from '@angular/common';
import { injectQueryClient, provideQueryClient } from './inject-query-client';
import { isDevMode } from './util/is-dev-mode/is-dev-mode';
/**
 * Sets up providers necessary to enable TanStack Query functionality for Angular applications.
 *
 * Allows to configure a `QueryClient` and optional features such as developer tools.
 *
 * **Example - standalone**
 *
 * ```ts
 * import {
 *   provideTanStackQuery,
 *   QueryClient,
 * } from '@tanstack/angular-query-experimental'
 *
 * bootstrapApplication(AppComponent, {
 *   providers: [provideTanStackQuery(new QueryClient())],
 * })
 * ```
 *
 * **Example - NgModule-based**
 *
 * ```ts
 * import {
 *   provideTanStackQuery,
 *   QueryClient,
 * } from '@tanstack/angular-query-experimental'
 *
 * @NgModule({
 *   declarations: [AppComponent],
 *   imports: [BrowserModule],
 *   providers: [provideTanStackQuery(new QueryClient())],
 *   bootstrap: [AppComponent],
 * })
 * export class AppModule {}
 * ```
 *
 * You can also enable optional developer tools by adding `withDevtools`. By
 * default the tools will then be loaded when your app is in development mode.
 * ```ts
 * import {
 *   provideTanStackQuery,
 *   withDevtools
 *   QueryClient,
 * } from '@tanstack/angular-query-experimental'
 *
 * bootstrapApplication(AppComponent,
 *   {
 *     providers: [
 *       provideTanStackQuery(new QueryClient(), withDevtools())
 *     ]
 *   }
 * )
 * ```
 * @param queryClient - A `QueryClient` instance.
 * @param features - Optional features to configure additional Query functionality.
 * @returns A set of providers to set up TanStack Query.
 * @public
 * @see https://tanstack.com/query/v5/docs/framework/angular/quick-start
 * @see withDevtools
 */
export function provideTanStackQuery(queryClient, ...features) {
    return makeEnvironmentProviders([
        provideQueryClient(queryClient),
        {
            provide: ENVIRONMENT_INITIALIZER,
            multi: true,
            useValue: () => {
                queryClient.mount();
                // Unmount the query client on application destroy
                inject(DestroyRef).onDestroy(() => queryClient.unmount());
            },
        },
        features.map((feature) => feature.ɵproviders),
    ]);
}
/**
 * Sets up providers necessary to enable TanStack Query functionality for Angular applications.
 *
 * Allows to configure a `QueryClient`.
 * @param queryClient - A `QueryClient` instance.
 * @returns A set of providers to set up TanStack Query.
 * @public
 * @see https://tanstack.com/query/v5/docs/framework/angular/quick-start
 * @deprecated Use `provideTanStackQuery` instead.
 */
export function provideAngularQuery(queryClient) {
    return provideTanStackQuery(queryClient);
}
/**
 * Helper function to create an object that represents a Query feature.
 * @param kind -
 * @param providers -
 * @returns A Query feature.
 */
function queryFeature(kind, providers) {
    return { ɵkind: kind, ɵproviders: providers };
}
/**
 * Enables developer tools.
 *
 * **Example**
 *
 * ```ts
 * export const appConfig: ApplicationConfig = {
 *   providers: [
 *     provideTanStackQuery(new QueryClient(), withDevtools())
 *   ]
 * }
 * ```
 * By default the devtools will be loaded when Angular runs in development mode and rendered in `<body>`.
 *
 * If you need more control over when devtools are loaded, you can use the `loadDevtools` option. This is particularly useful if you want to load devtools based on environment configurations. For instance, you might have a test environment running in production mode but still require devtools to be available.
 *
 * If you need more control over where devtools are rendered, consider `injectDevtoolsPanel`. This allows rendering devtools inside your own devtools for example.
 * @param optionsFn - A function that returns `DevtoolsOptions`.
 * @returns A set of providers for use with `provideTanStackQuery`.
 * @public
 * @see {@link provideTanStackQuery}
 * @see {@link DevtoolsOptions}
 */
export function withDevtools(optionsFn) {
    let providers = [];
    if (!isDevMode() && !optionsFn) {
        providers = [];
    }
    else {
        providers = [
            {
                provide: ENVIRONMENT_INITIALIZER,
                multi: true,
                useFactory: () => {
                    if (!isPlatformBrowser(inject(PLATFORM_ID)))
                        return () => { };
                    const injector = inject(Injector);
                    const options = computed(() => runInInjectionContext(injector, () => optionsFn?.() ?? {}));
                    let devtools = null;
                    let el = null;
                    const shouldLoadToolsSignal = computed(() => {
                        const { loadDevtools } = options();
                        return typeof loadDevtools === 'boolean'
                            ? loadDevtools
                            : isDevMode();
                    });
                    const doc = inject(DOCUMENT);
                    const destroyRef = inject(DestroyRef);
                    const getResolvedQueryClient = () => {
                        const injectedClient = injectQueryClient({
                            optional: true,
                            injector,
                        });
                        const client = options().client ?? injectedClient;
                        if (!client) {
                            throw new Error('No QueryClient found');
                        }
                        return client;
                    };
                    const destroyDevtools = () => {
                        devtools?.unmount();
                        el?.remove();
                        devtools = null;
                    };
                    return () => effect(() => {
                        const shouldLoadTools = shouldLoadToolsSignal();
                        const { client, position, errorTypes, buttonPosition, initialIsOpen, } = options();
                        if (devtools && !shouldLoadTools) {
                            destroyDevtools();
                            return;
                        }
                        else if (devtools && shouldLoadTools) {
                            client && devtools.setClient(client);
                            position && devtools.setPosition(position);
                            errorTypes && devtools.setErrorTypes(errorTypes);
                            buttonPosition && devtools.setButtonPosition(buttonPosition);
                            initialIsOpen && devtools.setInitialIsOpen(initialIsOpen);
                            return;
                        }
                        else if (!shouldLoadTools) {
                            return;
                        }
                        el = doc.body.appendChild(document.createElement('div'));
                        el.classList.add('tsqd-parent-container');
                        import('@tanstack/query-devtools').then((queryDevtools) => runInInjectionContext(injector, () => {
                            devtools = new queryDevtools.TanstackQueryDevtools({
                                ...options(),
                                client: getResolvedQueryClient(),
                                queryFlavor: 'Angular Query',
                                version: '5',
                                onlineManager,
                            });
                            el && devtools.mount(el);
                            // Unmount the devtools on application destroy
                            destroyRef.onDestroy(destroyDevtools);
                        }));
                    });
                },
            },
        ];
    }
    return queryFeature('DeveloperTools', providers);
}
export const queryFeatures = ['DeveloperTools'];
